### ISSUE TITLE
Meter name vs namespace

### ISSUE THREAD

**Issue by dyladan** [open-telemetry/opentelemetry-specification#410]
From the call today and other issues, there seems to be a lot of confusion around meter names vs namespaces. Some solutions have been proposed to deal with this but there seems not to be consensus which solves everybody's problems. I am opening this discussion issue in the hopes that we can resolve it quickly and move on with actual spec updates.

Here are the solutions that have been proposed that I am aware of:

### Use meter name as namespace

This is appealing because it completely removes the concept of namespace as a separate thing. It is on library developers to make sure that metric names within their library are unique, and if two libraries choose the same name, like "latency", then they are unique by virtue of being registered by separate libraries.

Examples:
- Library `mongodb` registers `latency`. Produced metrics are `namespace=mongodb, name=latency`
- Library `postgresql` registers `latency`. Produced metrics are `namespace=postgresql, name=latency`

Problems:
- If two libraries (mongodb-core and mongodb-native) both export latency and you want them to show up as a single latency metric. Perhaps this can be solved on the metrics backend?
- You are migrating from one library to another and you want the metrics provided by the new library to show up as the same metric as the old library. Perhaps this can be solved in the metrics backend as above?

### Require metric names to be fully qualified

In this scenario, it is entirely on library developers to give their metrics a name which is globally unique. This simplifies implementation, but could be problematic if a library developer decides to use a name that is too general or misunderstands the way the system works.

Examples:
- Library `mongodb` registers `mongodb_latency`
- Library `postgresql` registers `postgresql.latency`

Problems:
- Two libraries try to register the same name (mongo and psql both try to register `db_latency`)

### Use meter name as namespace, but allow it to be overridden by the operator

In this scenario, it is again up to library developers to make their metric names unique only within their library, but allows operators to override the namespace used by a given library for one reason or another.

Examples:
- Library `mongodb` registers `latency`. Produced metrics are `namespace=mongodb, name=latency`
- Library `mongodb-core` and `mongodb-native` both register `latency`. Produced metrics are `namespace=mongodb-core, name=latency` and `namespace=mongodb-native, name=latency` respectively.
	- The user overrides the namespace of both libraries to be `mongodb`. Produced metrics are now `namespace=mongodb, name=latency` for both of these libraries.

Problems:
- If two libraries each export a counter named `count`, and the user configures them to have the same namespace, does the meter need to return the same counter for both of them? If one of them increments 5 times and the other increments 3, is the value of the `count` counter 3, 5, or 8?

Example mock configuration:

```typescript
// operator initializes OT SDK
opentelemetry.initGlobalMeterProvider(
  new MeterProvider({
    namespaceOverrides: [
      {
        meterName: 'mongodb-native',
        namespace: 'mongodb'
      },
      {
        meterName: 'mongodb-core',
        namespace: 'mongodb'
      }
    ],
  })
);

// In library named mongodb-core
const coreMeter = opentelemetry.getMeter("mongodb-core", "3.2.2");
const coreLatency = coreMeter.createMeasure("latency");

// In library named mongodb-native
const nativeMeter = opentelemetry.getMeter("mongodb-native", "1.2.0");
const nativeLatency = nativeMeter.createMeasure("latency");

// In library named postgresql
const pgMeter = opentelemetry.getMeter("postgresql", "5.2.4");
const pgLatency = pgMeter.createMeasure("latency");

// The `namespace` property may not actually be exported, but is
// included here for illustrative purposes.
console.log(coreMeter.namespace); // mongodb
console.log(nativeMeter.namespace); // mongodb
console.log(pgMeter.namespace); // postgresql

/**
 * In this example, the two mongodb libraries have been configured
 * to have the same namespace by the operator. If the operator
 * had not configured namespace overrides, they would have had their own
 * separate namespaces.
 */
```

**Comment by dyladan** [comment]
ping @bogdandrutu @jmacd @danielkhan thoughts please :)

**Comment by arminru** [comment]
I don't quite understand the example where two libraries simultaneously report the exact same metric. When the operator overrides the namespace in your last example to `namespace=mongodb, name=latency`, would you expect two metrics with the same id to show up in the backend or would you expect them to be merged into one?
Or are the metrics reported by `mongodb-core` and `mongodb-native` actually different from each other since one is the latency observed by the wrapper library "core" and one the (lower) latency observed by the native component called by "core"? Then I would treat them as two distinct metrics in the backend. They could of course be charted in the same chart but are still two different metrics.

**Comment by dyladan** [comment]
@arminru they are simply two competing mongodb implementations. One microservice may use core and another might use native.

**Comment by dyladan** [comment]
Closed as a dup of #391

**Comment by dyladan** [comment]
>would you expect two metrics with the same id to show up in the backend or would you expect them to be merged into one?

I would probably expect them to be merged into one, but I am not a metrics expert and admit to not fully understanding the implications.
